mewarpx/mewarpx/emission.py

```diff
@@ -0,0 +1,868 @@
+            )
+
+            print("Inject {} particles".format(len(particles_dict['x'])))
+            # We can basically call _libwarpx.libwarpx.warpx_addNParticles()
+            # directly as we shouldn't need the wrapper, unless number to
+            # inject is 0, but for now we'll be conservative and use the
+            # wrapper.
+
+            # Note some parts of WarpX call the variables ux and some parts vx,
+            # and they're referred to as momenta. But I don't see anywhere
+            # they're actually used as momenta including the particle mass -
+            # the actual update is in Source/Particles/Pusher/UpdatePosition.H
+            _libwarpx.add_particles(
+                species_number=self.species.species_number,
+                x=particles_dict['x'],
+                y=particles_dict['y'],
+                z=particles_dict['z'],
+                ux=particles_dict['vx'],
+                uy=particles_dict['vy'],
+                uz=particles_dict['vz'],
+                # E_total=particles_dict['E_total'],
+                attr=particles_dict['w'],
+                unique_particles=self.unique_particles
+            )
+
+            if self.injector_diag is not None:
+                self.record_injectedparticles(
+                    self.species,
+                    particles_dict['E_total'],
+                    particles_dict.get('w', None)
+                )
+
+    def cleanup(self):
+        """Uninstall particle injection function
+        Previously used for testing to put Warp into a good state; could also
+        used if simulation should change partway through.
+        """
+        callbacks.uninstallparticleinjection(self.inject_particles)
+
+
+class ThermionicInjector(Injector):
+
+    """Performs standard every-timestep injection from a thermionic cathode."""
+
+    def __init__(self, emitter, species, npart_per_cellstep, T,
+                 WF, A=constants.A0*1e4,
+                 allow_poisson=False, wfac=1.0,
+                 name=None, profile_decorator=None,
+                 unique_particles=True):
+        """Sets up user-specified injection for warpX.
+        Arguments:
+            emitter (:class:`mewarpx.emission.Emitter`): Emitter object that
+                will specify positions and velocities of particles to inject.
+            species (mepicmi.Species): A premade species. Note only electrons
+                will actually give physically meaningful weight calculations.
+            npart_per_cellstep (int): Number of macroparticles to inject per
+                cell on the cathode surface per timestep
+            T (float): Cathode temperature (K)
+            WF (float): Cathode work function (eV)
+            A (float): Coefficient of emission in Amp/m^2/K^2. Default 1.2e6.
+            allow_poisson (bool): If True and < npart_per_cellstep electrons
+                would be injected per cell; inject whole electrons with a
```

Comment:
```suggestion
            allow_poisson (bool): If True and < npart_per_cellstep electrons
                would be injected per cell, inject whole electrons with a
```
The first part of this docstring for this argument isn't easy to understand on first read through. Specifically, the "< npart_per_cellstep electrons would be injected per cell" confuses me. Also the semicolon should be a comma to follow conditional statement syntax.